### PR TITLE
Chase and Attack for Melee and Ranged Enemies, Damage and Death, Garage Prototype

### PR DESCRIPTION
**Pull request by rob1997**
Attack, Chase, Damage and Death for Player and Enemies (Melee and Ranged)

### CODE REVIEW

**Review by sneakzttv — APPROVED** [review]

File: Assets/Scripts/NPC/Enemies/EnemyController.cs [hunk 1]
```diff
@@ -1,48 +1,28 @@
 using Chainsafe.BlockBlasterz.Events;
+using Chainsafe.BlockBlasterz.Weapons;
 using UnityEngine;
 
 namespace Chainsafe.BlockBlasterz
 {
-    public class EnemyController : NpcController
+    public abstract class EnemyController<T> : NpcController where T : WeaponBase
```
Comment (kantagara) [inline Assets/Scripts/NPC/Enemies/EnemyController.cs:7]: Why was enemys weapon tied to the controller itself? Couldnt it have its own Weapon Controller or something like that? I mean it is okay to do it like this but it feels a bit off 😅

File: Assets/Scripts/NPC/States/PatrolState.cs [hunk 2]
```diff
@@ -3,9 +3,9 @@
 
 namespace Chainsafe.BlockBlasterz
 {
-    public class PatrolState : State<PatrolState>
+    public class PatrolState : State<PatrolState, NpcController>
```
Comment (kantagara) [inline Assets/Scripts/NPC/States/PatrolState.cs:6]: okay even tho we have two generics, this is clever nevertheless 😄

File: Assets/Scripts/Player/PlayerPosition.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public struct PlayerSlot
+    {
+        [field: SerializeField] public Transform Slot { get; private set; }
+    }
+    
+    public class PlayerPosition : MonoBehaviour
+    {
+        public static PlayerPosition Instance { get; private set; }
+        
+        [field: SerializeField] public PlayerSlot[] Slots { get; private set; }
```
Comment (kantagara) [inline Assets/Scripts/Player/PlayerPosition.cs:18]: Why wasnt this just a simple list/array of transforms?

File: Assets/Scripts/Weapons/MeleeWeaponBase.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Chainsafe.BlockBlasterz.Weapons;
+using UnityEngine;
+using UnityEngine.Serialization;
+
+namespace Chainsafe.BlockBlasterz
+{
+    public class MeleeWeaponBase : WeaponBase
```
Comment (kantagara) [inline Assets/Scripts/Weapons/MeleeWeaponBase.cs:11]: Nice!

**Review by kantagara — APPROVED** [review]

File: Assets/Scripts/NPC/Enemies/EnemyController.cs (hunk 1 above)
Comment (rob1997) [reply Assets/Scripts/NPC/Enemies/EnemyController.cs:7]: so we can access weapons easily from States without casting, this is an abstract class we got two controllers inheriting, `RangedEnemyController` and `MeleeEnemyController`

File: Assets/Scripts/Player/PlayerPosition.cs (hunk 3 above)
Comment (rob1997) [reply Assets/Scripts/Player/PlayerPosition.cs:18]: so we can know when slot is occupied or not, currently working on this for #41

**Pull request merged**